src/standardizePronunciation/index.ts

```diff
@@ -0,0 +1,147 @@
+import { joinString } from '../_internal';
+import { isHangulAlphabet, isHangulCharacter } from '../_internal/hangul';
+import { combineHangulCharacter } from '../combineHangulCharacter';
+import { disassembleCompleteHangulCharacter } from '../disassembleCompleteHangulCharacter';
+import { isNotUndefined } from '../utils';
+import {
+  Nullable,
+  applyㄴㄹ덧남,
+  apply경음화,
+  apply제12항,
+  apply제13과14항,
+  apply제16항,
+  apply제17항,
+  apply제18항,
+  apply제19항,
+  apply제20항,
+  apply제9와10과11항,
+  type Syllable,
+} from './standardizePronunciation.rules';
+
+type NotHangul = {
+  index: number;
+  syllable: string;
+};
+
+/**
+ * 주어진 한글 문자열을 표준 발음으로 변환합니다.
+ * @param hangul 한글 문자열을 입력합니다.
+ * @param options 변환 옵션을 설정합니다.
+ * @param options.hardConversion 경음화 등의 된소리를 적용할지 여부를 설정합니다. 기본값은 true입니다.
+ * @returns 변환된 표준 발음 문자열을 반환합니다.
+ */
+export function standardizePronunciation(
+  hangul: string,
+  options: {
+    hardConversion: boolean;
+  } = { hardConversion: true }
+): string {
+  if (!hangul) {
+    return '';
+  }
+
+  const hangulPhrases = hangul.split(' ');
+  const changedHangul: string[] = [];
+
+  for (const hangulPhrase of hangulPhrases) {
+    const { notHangulPhrase, disassembleHangul } = 음절분해(hangulPhrase);
+
+    for (let i = 0; i < disassembleHangul.length; i += 1) {
+      const currentSyllable = disassembleHangul[i];
+      const nextSyllable = i < disassembleHangul.length - 1 ? disassembleHangul[i + 1] : null;
+
+      applyRules(currentSyllable, nextSyllable, i, hangulPhrase, options);
+    }
+
+    changedHangul.push(assembleChangedHangul(disassembleHangul, notHangulPhrase));
+  }
+
+  return changedHangul.join(' ');
```

Comment:
이 코드의 개선점이 2가지 생각나는데요, 

1. 전역 변수에 값이 반복문을 돌면서 계속 바뀐다,
2. 명령형 흐름으로 코드가 작성되어있어서, 인지적 부담이 있다.

전역 변수를 사용하지 않고, SRP원칙에 따라서 함수를 적절한 기준으로 쪼개보면 다음과 같이 사용할 수 있을 것 같아요. 

```suggestion
  const processSyllables = (syllables: string[], phrase: string, options: any) =>
    syllables.map((currentSyllable, I, array) => {
      const nextSyllable = I < array.length - 1 ? array[I + 1] : null;
      applyRules(currentSyllable, nextSyllable, I, phrase, options);
      return currentSyllable;
    });

  const transformHangulPhrase = (phrase: string, options: any): string => {
    const { notHangulPhrase, disassembleHangul } = 음절분해(phrase);
    const processedSyllables = processSyllables(disassembleHangul, phrase, options);
    return assembleChangedHangul(processedSyllables, notHangulPhrase);
  };

  return hangul
    .split(' ')
    .map(phrase => transformHangulPhrase(phrase, options))
    .join(' ');

```

Reply:
apply 쪽 코드들도, 대부분 명령형으로 코드가 구현되어있어서 코드를 읽는데에 시간이 많이 드는것 같아요.
위에서부터 아래로 순차적으로 눈을 내려가며 읽을걸 고려한다면, 조금 더 다른방향으로 리팩토링했으면 좋겠어요

Reply:
@okinawaa 위에서 언급하셨던 RORO 패턴과 순수함수 원칙을 적용했어요. 그리고 말씀주신 것처럼 apply 코드는 읽기에 쉽지 않다고 생각해서 각 규칙을 분리하여 선언형으로 작성하고 테스트 케이스를 추가했습니다.